Fix attribute rendering for boolean values (take 2) (#11660)
withastro/astro: .changeset/small-ties-sort.md

---
'astro': major
---

> Fixes attribute rendering for non-[boolean attributes](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML) with boolean values. This matches proper attribute handling in browsers and allows you to read the attribute values consistently.

Author: I think the context here is how the booleans are being rendered than being evaluated. Boolean attributes with boolean values are rendered as `<p attr>` or `<p>`. Non-boolean attributes with boolean attributes are rendered as `<p attr="true">` and `<p attr="false">`

I'm not quite sure how to re-word it from your suggestion, maybe we could remove the "read values consistently" part altogether if it's confusing?

> Fixes attribute rendering for non-[boolean attributes](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML) with boolean values. This matches proper attribute handling in browsers and allows you to read the attribute values consistently.

Reviewer: I do like the idea that "now you actually get the values rendered"...

What about something like:

Previously, non-boolean attributes may not have included their values when rendered to HTML. In Astro v.50...
???

> Fixes attribute rendering for non-[boolean attributes](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML) with boolean values. This matches proper attribute handling in browsers and allows you to read the attribute values consistently.

Author: Let's go with this then 👍 

```suggestion
Fixes attribute rendering for non-[boolean HTML attributes](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML) with boolean values to match proper attribute handling in browsers.

Previously, non-boolean HTML attributes may not have included their values when rendered to HTML. In Astro v5.0, the values are now explicitly rendered as `="true"` or `="false"`
```

Given this Astro code (where `allowfullscreen` is a boolean attribute):

```astro
<p allowfullscren={true}></p>
<p allowfullscren={false}></p>

<p unknown={true}></p>
<p unknown={false}></p>

<p data-foo={true}></p>
<p data-foo={false}></p>

> ```

Author: This looks great! Just checking since this is an Astro file and in the HTML-ish part, should the comments being in `<!-- -->` form instead? Or is the JS form the norm. I can update if so.

Thanks for the better names 😄

> The rendered HTML will be:

Reviewer: ```suggestion
Astro v5.0 now preserves the full data attribute with its value when rendering the HTML of non-boolean attributes:
```

```diff
  <p allowfullscren></p>
  <p></p>

  <p unknown="true"></p>
- <p unknown></p>
+ <p unknown="false"></p>

- <p data-foo></p>
+ <p data-foo="true"></p>
- <p></p>
+ <p data-foo="false"></p>
```

> To migrate, make sure you're reading the attribute values state correctly. For example:

Reviewer: ```suggestion
If you rely on attribute values, for example to locate elements or to conditionally render, update your code to match the new non-boolean attribute values:
```
Something like this maybe?

```diff

> - el.getAttribute('unknown') === ''
> + el.getAttribute('unknown') === 'false'
> 
> - el.hasAttribute('data-foo')
> + el.dataset.foo === 'true'

Reviewer: ```suggestion
- el.getAttribute('inherit') === ''
+ el.getAttribute('inherit') === 'false'

- el.hasAttribute('data-light')
+ el.dataset.foo === 'true'
```

---
'astro': major
---

Fixes attribute rendering for non-[boolean HTML attributes](https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML) with boolean values to match proper attribute handling in browsers.

Previously, non-boolean attributes may not have included their values when rendered to HTML. In Astro v5.0, the values are now explicitly rendered as `="true"` or `="false"`

> In the following `.astro` examples passed values of `true` and `false`, only `allowfullscreen` is a boolean attribute:

Reviewer: ```suggestion
In the following `.astro` examples, only `allowfullscreen` is a boolean attribute:
```
I know it was my line, but I kept stumbling over it every time I read it. It's probably good enough just to leave it out!

```astro
<!-- src/pages/index.astro -->
<!-- `allowfullscreen` is a boolean attribute -->
<p allowfullscreen={true}></p>
<p allowfullscreen={false}></p>

<!-- `inherit` is *not* a boolean attribute -->
<p inherit={true}></p>
<p inherit={false}></p>

<!-- `data-*` attributes are not boolean attributes -->
<p data-light={true}></p>
<p data-light={false}></p>
```

Astro v5.0 now preserves the full data attribute with its value when rendering the HTML of non-boolean attributes:

```diff

>   <p allowfullscren></p>

Reviewer 2: A tiny typo:
```suggestion
  <p allowfullscreen></p>
```

  <p></p>

>   <p unknown="true"></p>
> - <p unknown></p>
> + <p unknown="false"></p>
> 
> - <p data-foo></p>
> + <p data-foo="true"></p>
> - <p></p>
> + <p data-foo="false"></p>

Reviewer 2: Since you have renamed the attributes for lines 13 to 23 [in this review comment](https://github.com/withastro/astro/pull/11660#discussion_r1711519745), I think we have an inconsistency. This should be:
```suggestion
  <p inherit="true"></p>
- <p inherit></p>
+ <p inherit="false"></p>

- <p data-light></p>
+ <p data-light="true"></p>
- <p></p>
+ <p data-light="false"></p>
```

```

If you rely on attribute values, for example to locate elements or to conditionally render, update your code to match the new non-boolean attribute values:

```diff
- el.getAttribute('inherit') === ''
+ el.getAttribute('inherit') === 'false'

> - el.hasAttribute('data-light')
> + el.dataset.foo === 'true'

Reviewer 2: Same here, I think we have an inconsistency. This should be:
```suggestion
- el.hasAttribute('data-light')
+ el.dataset.light === 'true'
```
